ops/testing.py

```diff
@@ -149,6 +153,38 @@ def _create_meta(self, charm_metadata, action_metadata):
 
         return charm.CharmMeta.from_yaml(charm_metadata, action_metadata)
 
+    def _create_resources_dir(self) -> pathlib.Path:
+        """Create a temporary resource folder."""
+        self.__resource_dir = tempfile.TemporaryDirectory()
+        atexit.register(self.__resource_dir.cleanup)
+        return pathlib.Path(self.__resource_dir.name)
+
+    def add_oci_resource(self, resource_name: str = None, contents: typing.Mapping = None) -> None:
+        """Add oci resources to the backend.
+
+        This will register oci resources and create a temporary file for processing metadata
+        about the resource. A default set of values will be used for all oci resources unless
+        a specifc resource name and contents dict is proivded.
+
+        Args:
+            resource_name: Optioal name of the resource to add custom contents to.
+            contents: Optional custom dict to write for the named resource.
+        """
+        if resource_name and contents:
+            if self._meta.resources[resource_name].type == "oci-image":
+                resource_file = self._resource_dir / "{}".format(resource_name)
+                with resource_file.open('w') as resource_yaml:
+                    yaml.dump(contents, resource_yaml)
```

Comment:
So the actual path that Juju uses is
RESOURCE_DIR/<resource-name>/contents.yaml

While charms shouldn't depend on the exact path (as that is why resource-get returns a path), it might be reasonable to follow that pattern.


Also, this code silently fails if the type isn't "oci-image". I think we should just raise a RuntimeError if you try to add_oci_resource of a resource that isn't type oci-image.